### PR TITLE
Lazy evaluate Doorkeeper config when loading files and executing initializers

### PR DESCRIPTION
**Pull request by nbulaj**
Root of the issue is the fact that Rails or some specific gems (like Sorbet) could load model files **before** initializers have been run. Since some models have constructions like:

```ruby
class Some
  if Doorkeper.config.blabla
    # do smth
  end
end
```

in their class body - it will trigger an error because `@config` still doesn't exist (`Doorkeeper.configure` block wasn't called).

The same is applied to Doorkeeper railtie (engine): it adds some custom parameters filters based on the OAuth configuration, but it does it even for the new Rails apps which have not yet generated an initializer (just added a gem and run `rails g doorkeeper:install`).

What should be done:

* Avoid checking config values in railtie initializers (config could be nil for Rails apps which just installed the gem. see above issues)
* Avoid checking config in class definitions (move adding Polymorphic resource owner into an ORM hook)
* 
---

Addresses:

- https://github.com/doorkeeper-gem/doorkeeper/pull/1615
- https://github.com/doorkeeper-gem/doorkeeper/issues/1617
- https://github.com/doorkeeper-gem/doorkeeper/issues/1577

### CODE REVIEW

File: lib/doorkeeper/engine.rb [hunk 1]
```diff
@@ -2,10 +2,12 @@
 
 module Doorkeeper
   class Engine < Rails::Engine
-    initializer "doorkeeper.params.filter" do |app|
-      parameters = %w[client_secret authentication_token access_token refresh_token]
-      parameters << "code" if Doorkeeper.config.grant_flows.include?("authorization_code")
-      app.config.filter_parameters << /^(#{Regexp.union(parameters)})$/
+    initializer "doorkeeper.params.filter", after: :load_config_initializers  do |app|
```
Comment (johnnyshields) [inline lib/doorkeeper/engine.rb:5]: double whitespace before `do` here

File: lib/doorkeeper/config.rb [hunk 2]
```diff
@@ -33,6 +35,10 @@ def configuration
       @config || (raise MissingConfiguration)
```
Comment (johnnyshields) [inline lib/doorkeeper/config.rb:35]: This line still raises an error.

**Comment by zavan** [comment]
This broke my initializer in development (I haven't tested in production):

```
Jan 30 12:08:52 ubuntu-jammy rails[186539]: /usr/local/lib/ruby/gems/3.2.0/gems/activesupport-7.0.4.2/lib/active_support/inflector/methods.rb:280:in `constantize': uninitialized constant DoorkeeperAccessToken (NameError)
Jan 30 12:08:52 ubuntu-jammy rails[186539]:       Object.const_get(camel_cased_word)
Jan 30 12:08:52 ubuntu-jammy rails[186539]:             ^^^^^^^^^^
Jan 30 12:08:52 ubuntu-jammy rails[186539]:         from /usr/local/lib/ruby/gems/3.2.0/gems/activesupport-7.0.4.2/lib/active_support/core_ext/string/inflections.rb:74:in `constantize'
Jan 30 12:08:52 ubuntu-jammy rails[186539]:         from /usr/local/lib/ruby/gems/3.2.0/gems/doorkeeper-5.6.3/lib/doorkeeper/config.rb:425:in `access_token_model'
```

I have in my initializer/config:
```ruby
Doorkeeper.configure do
  # ...
  access_token_class 'DoorkeeperAccessToken'
  # ...
end
```

Rails v7.0.4.2

**Comment by nbulaj** [comment]
Yeah, can I see your `DoorkeeperAccessToken` implementation? Where it is located? How it looks like @zavan ?

**Comment by zavan** [comment]
> Yeah, can I see your `DoorkeeperAccessToken` implementation? Where it is located? How it looks like @zavan ?

It's a regular rails model in `app/models/doorkeeper_access_token.rb`:

```ruby
class DoorkeeperAccessToken < ApplicationRecord
  include ::Doorkeeper::Orm::ActiveRecord::Mixins::AccessToken

  after_save :sync_legacy_token

  private

  def sync_legacy_token
    #
  end
end
```

Downgrading to v5.6.2 works.

**Comment by nbulaj** [comment]
@zavan could you please try with `fix-custom-models` branch ?

**Comment by zavan** [comment]
> @zavan could you please try with `fix-custom-models` branch ?

It works 😄

**Pull request merged**